pandas/core/frame.py

```diff
@@ -4196,9 +4199,25 @@ def _maybe_casted_values(index, labels=None):
                     values.fill(np.nan)
                 else:
                     values = values.take(labels)
+
+                    # TODO(https://github.com/pandas-dev/pandas/issues/24206)
```

Comment:
grr this is messy

Reply:
I spent about 20 minutes on a proper solution, but getting proper support for EAs in `maybe_upcast_putmask` proved challenging.

Reply:
yes myabe_upcast_putmask is messy currently